Autotools cleanup rebased

fscheiner opened the pull request:
from @msalle:

This is a first step towards fixing #2 although it does not fix any of the points there, it does make building it a lot easier.

- add missing files AUTHORS, INSTALL, NEWS
- rename existing copyright and Changelog to autotools expected names.
- remove (ancient) output files from autotools: Makefile.in, aclocal.m4,
  compile, config.h.in, configure, depcomp, install-sh, missing
- remove GPT files pkg_data_src.gpt and pkg_data_src.gpt.in
- remove filelist, probably also from GPT?
- add ChangeLog entry for "no-longer-delete" patch
- few improvements in configure.ac:
    - update version, author and email
    - add check for C-compiler and preprocessor. The latter is required since
      $CPP is being used.
    - need to put errormsg between [] or a comma is parsed
    - update the default globus locations for source and EPEL
    - include both /lib and /lib64 for the LDFLAGS.
    - fix AC_INIT: 2.9-Beta needs hyphen, otherwise or make dist fails
    - add foreign to AM_INIT_AUTOMAKE to ease autotools
 - add EXTRA_DIST to Makefile.am to make sure we get all files.

_**commit 23-7-2020**_
- added support for pkg-config, CPPFLAGS and LDFLAGS.
- Only CFLAGS is not used.

msalle committed:
Cleanup/fix building, update for gridcf

- add missing files AUTHORS, INSTALL, NEWS
- rename existing copyright and Changelog to autotools expected names.
- remove (ancient) output files from autotools: Makefile.in, aclocal.m4,
  compile, config.h.in, configure, depcomp, install-sh, missing
- remove GPT files pkg_data_src.gpt and pkg_data_src.gpt.in
- remove filelist, probably also from GPT?
- add ChangeLog entry for "no-longer-delete" patch
- few improvements in configure.ac:
    - update version, author and email
    - add check for C-compiler and preprocessor. The latter is required since
      $CPP is being used.
    - need to put errormsg between [] or a comma is parsed
    - update the default globus locations for source and EPEL
    - include both /lib and /lib64 for the LDFLAGS.
- add EXTRA_DIST to Makefile.am to make sure we get all files.
- fix AC_INIT: 2.9-Beta needs hyphen, otherwise or make dist fails
- add foreign to AM_INIT_AUTOMAKE to easy autotools

Ideally the check for the globus include and libs should (also) be done with
pkgconfig macros, but this should at least get it to build again from both
source installer *and* EPEL. Also it does not yet honour CFLAGS, LDFLAGS etc.

fscheiner committed:
Update ChangeLog

fscheiner commented on configure.ac:
```diff
@@ -118,18 +122,13 @@ AC_ARG_WITH(globus,
 	# RPM installation
-	if test "$FLAVOR" = "32"; then
-		CPPFLAGS=-I$globus_location/lib/globus/include
-	else
-		CPPFLAGS=-I$globus_location/lib64/globus/include
-	fi
-
+	CPPFLAGS=-I$globus_location/include/globus
```
Just to be sure: Is this (and the similar change below) also supposed to work for 32 bit builds?

msalle replied:
For the CPPFLAGS, which set here the include flags as used for RPMs, this should no longer be arch dependent. Only the LDFLAGS are still arch dependent (obviously) and there we simply check both, first 64bit then 32bit. Ultimately we need to synchronise this with the rest of the GCT build process.

fscheiner replied:
OK then.

msalle commented:
Hi all, I've now properly ff-ed my autotools_cleanup_rebased branch (couldn't do that beforehand), such that it itself is just a ff from the master branch. I fixed one more thing in the configure.ac (adding foreign to AM_INIT_AUTOMAKE).
I think it's ready now.

fscheiner commented on INSTALL:
```diff
@@ -0,0 +1,368 @@
+Basic Installation
+==================
+
+   Briefly, the shell command './configure && make && make install'
```
@msalle: Didn't notice this until I tried to build UberFTP manually on openSUSE Leap 15.1: `configure` is removed with this patch set, only `configure.ac` remains. I think this is similar to the GCT, but then this INSTALL document is missing the step to actually create `configure` (e.g. `autoreconf -i`?).

msalle replied:
This is the normal setup with the autotools in my opinion: you don't want to include the output of the autotools in the git or svn repository, but you do want to have an INSTALL file, which will end up in the output of "make dist", which is what you're distributing. If you leave out the INSTALL file, some of the autotools complain.
We could add an `autogen.sh` or `bootstrap` script, similar to those of e.g. [VOMS](https://github.com/italiangrid/voms/blob/master/autogen.sh) or [LCMAPS](https://ndpfsvn.nikhef.nl/viewvc/mwsec/trunk/lcmaps/bootstrap?view=markup)

fscheiner replied:
Ok, I agree, though I seem to remember seeing other software distributing a ready-made `configure` script, but these possibly don't use autotools. [The autotools FAQ](https://www.gnu.org/software/automake/faq/autotools-faq.html#What-does-_002e_002fbootstrap-or-_002e_002fautogen_002esh-do_003f) recommends to use `bootstrap` over `autogen.sh`. For me `autoreconf -i` was enough to get things ready for the configure run. Do we need anything in addition in this file?

msalle replied:
Probably not, `autoreconf -i` is basically doing all those steps (I would then add the `-v` as in the FAQ you quoted). That's also why many people don't bother shipping a `bootstrap` file I think.

fscheiner replied:
Ok, looks like I cannot create another commit for this PR to add a `bootstrap`- most likely because I created it using a branch from your fork instead of letting you create it from your fork - otherwise it could allow me to make changes in any remote branch on GitHub I take as base for a PR. I'd propose the following contents for `bootstrap`:

```
#!/bin/sh
autoreconf -vi
```

...if you like also with `space` between shebang and path, but I prefer it w/o.

msalle replied:
I've just added a bootstrap.

fscheiner replied:
Ok, great.

msalle committed:
Add support pkg-config, CPPFLAGS, LDFLAGS

Add missing support for pkg-config, and for using environment variables CPPFLAGS
and LDFLAGS.
We still ignore CFLAGS which is less useful.

msalle committed:
Adding very simple bootstrap

Running autoreconf is sufficient to get all the autotools to work.

msalle commented:
I've just added two more commits: one two add a `bootstrap` and the second to add support for the `pkg-config` plus `CPPFLAGS` and `LDFLAGS` environment variables.

fscheiner commented:
I just tested a `uberftp` compiled from the sources with this patch set applied and directly ran it from the sources directory and it behaves strange: When doing an `ls` on our GridFTP server, the server shows that it sent the directory listing, but `uberftp` does not show anything and also does not return. Hitting `Ctrl+c` ends `uberftp` and also closes the connection. The older v2.8 `uberftp` I created from the EPEL/Fedora sources (available from https://build.opensuse.org/package/show/home:frank_scheiner:gct/uberftp) does work as expected in this regard. Does it (v2.9-Beta) work for you?

I'm currently struggling to build an `uberftp` without this patch set - the configure run can't find `globus_config.h` for some reason - so can't really tell if this patch set is responsible. But from the changes made here I assume the problems might be coming from an older commit.

fscheiner commented:
Ok, found the culprit: some paths in `configure(.ac)` were wrong for openSUSE Leap 15.1:
```
diff --git a/configure b/configure
index ccde4e3..539d1b6 100755
--- a/configure
+++ b/configure
@@ -4674,9 +4674,9 @@ fi
 if test "x$globus_location" == "x/usr" ; then
 	# RPM installation
 	if test "$FLAVOR" = "32"; then
-		CPPFLAGS=-I$globus_location/lib/globus/include
+		CPPFLAGS=-I$globus_location/include/globus
 	else
-		CPPFLAGS=-I$globus_location/lib64/globus/include
+		CPPFLAGS=-I$globus_location/include/globus
 	fi
 
 

```
Testing a `uberftp` compiled from the master branch gives the same result, so something between https://github.com/gridcf/UberFTP/commit/012788f5430c9f7eb03e65b7aa8bcb106f472518 and https://github.com/gridcf/UberFTP/commit/7bbe1ff19c27028ae3d142e4f1f98e162b95b610 is wrong. I should have tested these changes. :-/

fscheiner commented:
Ok, according to bisecting and my testing, uberftp is bad since https://github.com/gridcf/UberFTP/commit/6cd744627cbaadfc5797109060139b83cd1a2566 (corresponding PR: https://github.com/JasonAlt/UberFTP/pull/5). Building from https://github.com/gridcf/UberFTP/commit/012788f5430c9f7eb03e65b7aa8bcb106f472518 gives a working version (when adapting the paths in `configure.ac` and doing a `autoreconf -vi` prior to `./configure && make`), which is the state at which we forked UberFTP.

msalle commented:
> Ok, found the culprit: some paths in `configure(.ac)` were wrong for openSUSE Leap 15.1:
> 
> ```
> diff --git a/configure b/configure
> index ccde4e3..539d1b6 100755
> --- a/configure
> +++ b/configure
> @@ -4674,9 +4674,9 @@ fi
>  if test "x$globus_location" == "x/usr" ; then
>  	# RPM installation
>  	if test "$FLAVOR" = "32"; then
> -		CPPFLAGS=-I$globus_location/lib/globus/include
> +		CPPFLAGS=-I$globus_location/include/globus
>  	else
> -		CPPFLAGS=-I$globus_location/lib64/globus/include
> +		CPPFLAGS=-I$globus_location/include/globus
>  	fi
>  
>  
> ```

You mean this fixes it, or this is wrong? It looks like https://github.com/msalle/UberFTP/commit/2b7d81b41c1fe31911dd3bfe4b0532ca54e3d5be but differently? Note that the latter is also superceded by my commit from yesterday.
It's always a bit tricky: the (EPEL) globus RPMs install the headers in `/usr/lib(64)/include/globus`, but the source installer installs them directly in `PREFIX/include`. If you give a `--with-globus` for the RPMs this therefore doesn't work smoothly. For RPMs in default location you should not need that, and you can now also use pkg-config i.e. specify a `PKG_CONFIG_PATH`

msalle commented:
> I just tested a `uberftp` compiled from the sources with this patch set applied and directly ran it from the sources directory and it behaves strange: When doing an `ls` on our GridFTP server, the server shows that it sent the directory listing, but `uberftp` does not show anything and also does not return. Hitting `Ctrl+c` ends `uberftp` and also closes the connection. The older v2.8 `uberftp` I created from the EPEL/Fedora sources (available from https://build.opensuse.org/package/show/home:frank_scheiner:gct/uberftp) does work as expected in this regard. Does it (v2.9-Beta) work for you?
> 
> I'm currently struggling to build an `uberftp` without this patch set - the configure run can't find `globus_config.h` for some reason - so can't really tell if this patch set is responsible. But from the changes made here I assume the problems might be coming from an older commit.

I don't see this, I've just tested my source build against the GCT 6.2.1567772254 and 6.2.1541705016 built from source installer and both using `./uberftp prometheus.desy.de ls` and `./uberftp -ls gsiftp://prometheus.desy.de/` and it all just works?

fscheiner commented:
> > Ok, found the culprit: some paths in `configure(.ac)` were wrong for openSUSE Leap 15.1:
> > ```
> >  [...]
> > ```
> 
> You mean this fixes it, or this is wrong?

The diff fixes the issue for me.

> It looks like [msalle@2b7d81b](https://github.com/msalle/UberFTP/commit/2b7d81b41c1fe31911dd3bfe4b0532ca54e3d5be) but differently?

This was done based on https://github.com/gridcf/UberFTP/commit/7bbe1ff19c27028ae3d142e4f1f98e162b95b610 (i.e. w/o your patches applied).

> Note that the latter is also superceded by my commit from yesterday.

I believe my local copy was up to date (with your latest commit) when I did the bisecting.

> It's always a bit tricky: the (EPEL) globus RPMs install the headers in `/usr/lib(64)/include/globus`, but the source installer installs them directly in `PREFIX/include`. If you give a `--with-globus` for the RPMs this therefore doesn't work smoothly. For RPMs in default location you should not need that, and you can now also use pkg-config i.e. specify a `PKG_CONFIG_PATH`

Indeed.

fscheiner commented:
> > I just tested a `uberftp` compiled from the sources with this patch set applied and directly ran it from the sources directory and it behaves strange: [...]
> 
> I don't see this, I've just tested my source build against the GCT 6.2.1567772254 and 6.2.1541705016 built from source installer and both using `./uberftp prometheus.desy.de ls` and `./uberftp -ls gsiftp://prometheus.desy.de/` and it all just works?

Strange. :-| But do you also have a Globus GridFTP server to test against? I'll check against a local GridFTP server to be sure it's not related to our GridFTP server.

fscheiner commented:
> > > I just tested a `uberftp` compiled from the sources with this patch set applied and directly ran it from the sources directory and it behaves strange: [...]
> > 
> > 
> > I don't see this, I've just tested my source build against the GCT 6.2.1567772254 and 6.2.1541705016 built from source installer and both using `./uberftp prometheus.desy.de ls` and `./uberftp -ls gsiftp://prometheus.desy.de/` and it all just works?
> 
> Strange. :-| But do you also have a Globus GridFTP server to test against? I'll check against a local GridFTP server to be sure it's not related to our GridFTP server.

@msalle 
Ok, a uberftp built with this patchset works correctly with a local GridFTP server at my lab, but not with our GridFTP server at HLRS. Then I'd say, let's just pull these changes and put uberftp closer to v2.9.

fscheiner commented:
@msalle, @matyasselmeci, @ellert:
As I created that PR - actually to save us some time - I am not allowed to perform an approving review by the GitHub UI. :-(

So @matyasselmeci or @ellert, would you be so kind and approve (and merge) these changes. From my testing with a local Globus GridFTP server and @msalle's testing with dCache they are working well.

msalle commented:
> > > Ok, found the culprit: some paths in `configure(.ac)` were wrong for openSUSE Leap 15.1:
> > > ```
> > >  [...]
> > > ```
> > 
> > 
> > You mean this fixes it, or this is wrong?
> 
> The diff fixes the issue for me.
> 
> > It looks like [msalle@2b7d81b](https://github.com/msalle/UberFTP/commit/2b7d81b41c1fe31911dd3bfe4b0532ca54e3d5be) but differently?
> 
> This was done based on [7bbe1ff](https://github.com/gridcf/UberFTP/commit/7bbe1ff19c27028ae3d142e4f1f98e162b95b610) (i.e. w/o your patches applied).

you mean your diff fixes the build problems, which also should be fixed by my commits, but ...

> > Note that the latter is also superceded by my commit from yesterday.
> 
> I believe my local copy was up to date (with your latest commit) when I did the bisecting.

... my commit should not change anything about this run-time problem. So just to be clear: the failure to run must be caused by one of the commits we now have imported from the devel branch, hence this PR should not interfere with a potential solution of the other issue. Do you agree with that?

> > It's always a bit tricky: the (EPEL) globus RPMs install the headers in `/usr/lib(64)/include/globus`, but the source installer installs them directly in `PREFIX/include`. If you give a `--with-globus` for the RPMs this therefore doesn't work smoothly. For RPMs in default location you should not need that, and you can now also use pkg-config i.e. specify a `PKG_CONFIG_PATH`
> 
> Indeed.

just to make also this clear: do you have any build-problems when using my latest commits?

fscheiner commented:
> > > > Ok, found the culprit: some paths in `configure(.ac)` were wrong for openSUSE Leap 15.1:
> [...]
> 
> you mean your diff fixes the build problems, which also should be fixed by my commits

Yes. As I did a bisecting I had to checkout older revisions of the code, which needed some changes before I could configure and compile them on openSUSE Leap 15.1. This is not related to your changes, although they fix the same problem.

>, but ...
> 
> > > Note that the latter is also superceded by my commit from yesterday.
> > 
> > 
> > I believe my local copy was up to date (with your latest commit) when I did the bisecting.
> 
> ... my commit should not change anything about this run-time problem. So just to be clear: the failure to run must be caused by one of the commits we now have imported from the devel branch, hence this PR should not interfere with a potential solution of the other issue. Do you agree with that?

Yes, I totally agree. Sorry, I thought I had made that clear with https://github.com/gridcf/UberFTP/pull/8#issuecomment-663129115 already, but obviously not.

> 
> > > It's always a bit tricky: the (EPEL) globus RPMs install the headers in `/usr/lib(64)/include/globus`, but the source installer installs them directly in `PREFIX/include`. If you give a `--with-globus` for the RPMs this therefore doesn't work smoothly. For RPMs in default location you should not need that, and you can now also use pkg-config i.e. specify a `PKG_CONFIG_PATH`
> > 
> > 
> > Indeed.
> 
> just to make also this clear: do you have any build-problems when using my latest commits?

No.

fscheiner commented:
@ellert: Do you have time to perform a quick review of these changes? Because two other PRs are ready to be merged after rebased to the source with the changes from this PR included and this one's already open since end of July. Please see https://github.com/gridcf/UberFTP/pull/8#issuecomment-666363284 for why.

ellert reviewed (approved):
Looks good.
